recirq/otoc/loschmidt/tilted_square_lattice/analysis.py

```diff
@@ -0,0 +1,194 @@
+
+    Args:
+        df: The dataframe
+        y_cols: Columnn not to group by. These will be aggregated over.
+        agg_func: The aggregation function
+
+    Returns:
+        The dataframe and list of columns used in the groupby.
+    """
+    groupby_cols = list(df.columns)
+    for yc in y_cols:
+        try:
+            groupby_cols.remove(yc)
+        except ValueError as e:
+            raise ValueError(f"{yc} not in columns {groupby_cols}.")
+
+    agged = df.groupby(groupby_cols, as_index=False).agg(agg_func)
+
+    if isinstance(agged.columns, pd.MultiIndex):
+        agged.columns = ['_'.join(x for x in column_tuple if x)
+                         for column_tuple in agged.columns]
+
+    return agged, groupby_cols
+
+
+def _results_to_dataframe(
+        results: ExecutableGroupResult,
+        func: Callable[[ExecutableResult, QuantumRuntimeConfiguration, SharedRuntimeInfo], Dict]
+) -> pd.DataFrame:
+    """Call a function on each result in an `ExecutableGroupResult` to construct a DataFrame."""
+    return pd.DataFrame([func(result, results.runtime_configuration, results.shared_runtime_info)
+                         for result in results.executable_results])
+
+
+def loschmidt_results_to_dataframe(results: ExecutableGroupResult) -> pd.DataFrame:
+    """Process an `ExecutableGroupResult`.
+
+    This function performs the data analysis using `to_ground_state_prob` and
+    extracts the most relevant quantities for a dataframe suitable for
+    grouping, selecting, and plotting.
+    """
+
+    def _to_record(result: ExecutableResult,
+                   rt_config: QuantumRuntimeConfiguration,
+                   shared_rt_info: SharedRuntimeInfo) -> Dict:
+        success_prob = to_ground_state_prob(result.raw_data)
+        spec = cast(TiltedSquareLatticeLoschmidtSpec, result.spec)
+
+        return {
+            'run_id': shared_rt_info.run_id,
+            'width': spec.topology.width,
+            'height': spec.topology.height,
+            'n_qubits': spec.topology.n_nodes,
+            'macrocycle_depth': spec.macrocycle_depth,
+            'q_area': (spec.macrocycle_depth * 4 * 2) * spec.topology.n_nodes,
+            'instance_i': spec.instance_i,
+            'n_repetitions': spec.n_repetitions,
+            'success_probability': success_prob,
+            'processor_id': rt_config.processor_record.processor_id,
+        }
+
+    return _results_to_dataframe(results, _to_record)
+
+
+def fit_vs_macrocycle_depth(df):
+    """Perform a fitting vs macrocycle depth given a loschmidt results dataframe.
+
+    Args:
+        df: The dataframe from `loschmidt_results_to_dataframe`.
+
+    Returns:
+        fitted_df: A new dataframe containing fit parameters.
+        exp_ansatz_vs_macrocycle_depth: The function used for the fit.
+    """
+
+    def exp_ansatz_vs_macrocycle_depth(macrocycle_depths, a, f):
+        return a * np.exp(np.log(f) * 4 * 2 * macrocycle_depths)
```

Comment:
Is the meaning of 4 and 2 well known? Should they be named constants?

Reply:
added a beefy note to the docstring of `loschmidt_results_to_dataframe` and a shorter line comment

Reply:
and i'll factor out as named constants